cpp/src/arrow/table_test.cc

```diff
@@ -406,6 +407,163 @@ TEST_F(TestTable, ConcatenateTables) {
   ASSERT_RAISES(Invalid, ConcatenateTables({t1, t3}, &result));
 }
 
+using TestPromoteTableToSchema = TestTable;
+
+TEST_F(TestPromoteTableToSchema, IdenticalSchema) {
+  const int length = 10;
+  MakeExample1(length);
+  std::shared_ptr<Table> table = Table::Make(schema_, arrays_);
+
+  std::shared_ptr<Table> result;
+  auto metadata =
+      std::shared_ptr<KeyValueMetadata>(new KeyValueMetadata({"foo"}, {"bar"}));
+  ASSERT_OK(PromoteTableToSchema(default_memory_pool(), table,
+                                 schema_->WithMetadata(metadata), &result));
+
+  ASSERT_TRUE(table->ReplaceSchemaMetadata(metadata)->Equals(*result));
+}
+
+TEST_F(TestPromoteTableToSchema, PromoteNullTypeField) {
+  const int length = 10;
+  auto metadata =
+      std::shared_ptr<KeyValueMetadata>(new KeyValueMetadata({"foo"}, {"bar"}));
+  auto f0 = field("f0", null())->WithMetadata(metadata);
+  auto f0_alt = field("f0", int32());
+  auto table = Table::Make(schema({f0}), {MakeRandomArray<NullArray>(length)});
+
+  std::shared_ptr<Table> result;
+  ASSERT_OK(
+      PromoteTableToSchema(default_memory_pool(), table, schema({f0_alt}), &result));
+  std::shared_ptr<Array> array_of_nulls;
+  ASSERT_OK(MakeArrayOfNull(int32(), length, &array_of_nulls));
+  ASSERT_TRUE(result->Equals(*Table::Make(schema({f0_alt}), {array_of_nulls})));
+}
+
+TEST_F(TestPromoteTableToSchema, AddMissingField) {
+  const int length = 10;
+  auto f0 = field("f0", int32());
+  auto table = Table::Make(schema({}), std::vector<std::shared_ptr<Array>>(), length);
+  std::shared_ptr<Table> result;
+  ASSERT_OK(
+      PromoteTableToSchema(default_memory_pool(), table, schema({f0}), &result));
+
+  std::shared_ptr<Array> array_of_nulls;
+  ASSERT_OK(MakeArrayOfNull(int32(), length, &array_of_nulls));
+  ASSERT_TRUE(result->Equals(*Table::Make(schema({f0}), {array_of_nulls})));
+}
+
+TEST_F(TestPromoteTableToSchema, IncompatibleTypes) {
+  const int length = 10;
+
+  auto f0 = field("f0", int32());
+  auto f0_alt1 = field("f0", null());
+  auto f0_alt2 = field("f0", uint32());
+
+  auto table = Table::Make(schema({f0}), {MakeRandomArray<NullArray>(length)});
+
+  std::shared_ptr<Table> result;
+  Status s =
+      PromoteTableToSchema(default_memory_pool(), table, schema({f0_alt1}), &result);
+  ASSERT_TRUE(s.IsInvalid());
+
+  s = PromoteTableToSchema(default_memory_pool(), table, schema({f0_alt2}), &result);
+  ASSERT_TRUE(s.IsInvalid());
+}
+
+TEST_F(TestPromoteTableToSchema, DuplicateFieldNames) {
+  const int length = 10;
+
+  auto f0 = field("f0", int32());
+  auto f0_alt = field("f0", null());
+
+  auto table = Table::Make(schema({f0, f0_alt}), {MakeRandomArray<Int32Array>(length),
+                                                  MakeRandomArray<NullArray>(length)});
+
+  std::shared_ptr<Table> result;
+  Status s = PromoteTableToSchema(default_memory_pool(), table, schema({f0}), &result);
```

Comment:
Please use ASSERT_RAISES for errors
```suggestion
  ASSERT_RAISES(Invalid, PromoteTableToSchema(default_memory_pool(), table, schema({f0}), &result));
```